nixos/modules/services/web-apps/mastodon.nix

```diff
@@ -588,26 +588,184 @@ in {
+          tryFiles = "$uri =404";
+          priority = 2240;
+
+          extraConfig = ''
+            add_header Cache-Control 'public, max-age=2419200, must-revalidate';
+            add_header Strict-Transport-Security 'max-age=63072000; includeSubDomains';
+            ${lib.optionalString config.services.nginx.virtualHosts.${cfg.localDomain}.http3 ''
+              add_header Alt-Svc 'h3=":443"; ma=86400';
+            ''}
+          '';
+        };
+
+        locations."^~ /packs/" = {
+          tryFiles = "$uri =404";
+          priority = 2250;
+
+          extraConfig = ''
+            add_header Cache-Control 'public, max-age=2419200, must-revalidate';
+            add_header Strict-Transport-Security 'max-age=63072000; includeSubDomains';
+            ${lib.optionalString config.services.nginx.virtualHosts.${cfg.localDomain}.http3 ''
+              add_header Alt-Svc 'h3=":443"; ma=86400';
+            ''}
+          '';
         };
 
-        locations."/api/v1/streaming/" = {
+        locations."^~ /shortcuts/" = {
+          tryFiles = "$uri =404";
+          priority = 2260;
+
+          extraConfig = ''
+            add_header Cache-Control 'public, max-age=2419200, must-revalidate';
+            add_header Strict-Transport-Security 'max-age=63072000; includeSubDomains';
+            ${lib.optionalString config.services.nginx.virtualHosts.${cfg.localDomain}.http3 ''
+              add_header Alt-Svc 'h3=":443"; ma=86400';
+            ''}
+          '';
+        };
+
+        locations."^~ /sounds/" = {
+          tryFiles = "$uri =404";
+          priority = 2270;
+
+          extraConfig = ''
+            add_header Cache-Control 'public, max-age=2419200, must-revalidate';
+            add_header Strict-Transport-Security 'max-age=63072000; includeSubDomains';
+            ${lib.optionalString config.services.nginx.virtualHosts.${cfg.localDomain}.http3 ''
+              add_header Alt-Svc 'h3=":443"; ma=86400';
+            ''}
+          '';
+        };
+
+        locations."^~ /system/" = {
+          tryFiles = "$uri =404";
+          alias = "/var/lib/mastodon/public-system/";
+          priority = 2280;
+
+          extraConfig = ''
+            add_header Cache-Control "public, max-age=2419200, immutable";
+            add_header Strict-Transport-Security 'max-age=63072000; includeSubDomains';
+            ${lib.optionalString config.services.nginx.virtualHosts.${cfg.localDomain}.http3 ''
+              add_header Alt-Svc 'h3=":443"; ma=86400';
+            ''}
+          '';
+        };
+
+        locations."^~ /api/v1/streaming/" = {
           proxyPass = (if cfg.enableUnixSocket then "http://unix:/run/mastodon-streaming/streaming.socket" else "http://127.0.0.1:${toString(cfg.streamingPort)}/");
           proxyWebsockets = true;
+          priority = 2300;
+
+          extraConfig = ''
+            proxy_set_header Proxy "";
+
+            proxy_buffering off;
+            proxy_redirect off;
```

Comment:
Is nginx buffering websockets at all? Also upstream recommends against this unless long polling is used.

Reply:
I find it difficult to answer.
I have been using this configuration on my instance for a long time. I didn't notice any suspicious errors.

Reply:
It might be that errors only occur in very specific scenarios and not at all on lightly used instances. In question I'd rather would like to drop this.

Reply:
In upstream the `proxy_redirect` parameter is disabled:
```
  location ^~ /api/v1/streaming {
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Proxy "";

    proxy_pass http://streaming;
    proxy_buffering off;
    proxy_redirect off;
...
```

Reply:
To me this looks fine, as in it replicates what upstream does? That is the only thing I understand about this tbh, wether it replicates the behaviour recommended by upstream.

Reply:
I set the same parameters and values as in the upstream.
I don't know much about proxying.

Reply:
I am not feeling to great about just copying upstream values especially if we are not sure if they also not just copied that from somewhere without deeper thought.